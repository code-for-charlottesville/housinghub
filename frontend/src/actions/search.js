export function setPaginationQuery(fieldName, newValue) {
  return {
    type: "SET_PAGINATION_QUERY",
    fieldName,
    newValue,
  };
}

export function setSearchfieldsQuery(fieldName, newValue) {
  return {
    type: "SET_SEARCHFIELDS_QUERY",
    fieldName,
    newValue,
  };
}

export function setPaginationResults(fieldName, newValue) {
  return {
    type: "SET_PAGINATION_RESULTS",
    fieldName,
    newValue,
  };
}

export function setSearchResults(resultList) {
  return {
    type: "SET_SEARCH_RESULTS",
    resultList,
  };
}
