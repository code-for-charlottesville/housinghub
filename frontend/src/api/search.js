import axios from "axios";
import apiEndpoint from "./endpoint";

/**
 * makes POST request to /backend/property/search with query
 **/

export function postQuery(query) {
  console.log("postQuery");
  return axios
    .post(`${apiEndpoint}/property/search`, query)
    .then((r) => {
      return Promise.resolve(r.data);
    })
    .catch((err) => {
      return Promise.resolve(err.response.data);
    });
}
