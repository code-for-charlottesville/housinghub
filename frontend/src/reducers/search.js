const initialState = {
  errors: "",
  searchResults: {
    pagination: {
      page: 0,
      results_per_page: 0,
      totalNumberOfResults: 0,
    },
    results: [],
  },
  query: {
    pagination: {
      page: 0,
      results_per_page: 15,
    },
    searchFields: {},
  },
};

const searchState = (state = initialState, action) => {
  switch (action.type) {
    case "SET_PAGINATION_QUERY":
      return Object.assign({}, state, {
        ...state,
        query: {
          pagination: {
            ...state.query.pagination,
            [action.fieldName]: action.newValue,
          },
          searchFields: {
            ...state.query.searchFields,
          },
        },
      });
    case "SET_SEARCHFIELDS_QUERY":
      if (Array.isArray(state.query.searchFields[action.fieldName])) {
        let index = state.query.searchFields[action.fieldName].indexOf(
          action.newValue
        );
        if (index > -1) {
          return Object.assign({}, state, {
            ...state,
            query: {
              pagination: {
                ...state.query.pagination,
              },
              searchFields: {
                ...state.query.searchFields,
                [action.fieldName]: state.query.searchFields[
                  action.fieldName
                ].filter((item) => item !== action.newValue),
              },
            },
          });
        } else {
          return Object.assign({}, state, {
            ...state,
            query: {
              pagination: {
                ...state.query.pagination,
              },
              searchFields: {
                ...state.query.searchFields,
                [action.fieldName]: [
                  ...state.query.searchFields[action.fieldName],
                  action.newValue,
                ],
              },
            },
          });
        }
      } else {
        return Object.assign({}, state, {
          ...state,
          query: {
            pagination: {
              ...state.query.pagination,
            },
            searchFields: {
              ...state.query.searchFields,
              [action.fieldName]: action.newValue,
            },
          },
        });
      }
    case "SET_PAGINATION_RESULTS":
      return Object.assign({}, state, {
        ...state,
        searchResults: {
          ...state.searchResults,
          pagination: {
            ...state.searchResults.pagination,
            [action.fieldName]: action.newValue,
          },
        },
      });
    case "SET_SEARCH_RESULTS":
      return Object.assign({}, state, {
        ...state,
        searchResults: action.response,
      });
    default:
      return state;
  }
};

export default searchState;
