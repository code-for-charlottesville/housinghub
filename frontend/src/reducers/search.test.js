import search from "./search";

describe("reducers/appState", () => {
  let testTable = [
    {
      name: "set pagination query",
      action: {
        type: "SET_PAGINATION_QUERY",
        fieldName: "page",
        newValue: 3,
      },
    },
    {
      name: "set search fields query",
      action: {
        type: "SET_SEARCHFIELDS_QUERY",
        fieldName: "bathrooms",
        newValue: 2,
      },
    },
    {
      name: "set pagination results",
      action: {
        type: "SET_PAGINATION_RESULTS",
        fieldName: "page",
        newValue: 5,
      },
    },
    {
      name: "set query results",
      action: {
        type: "SET_SEARCH_RESULTS",
        fieldName: "bathrooms",
        resultList: [],
      },
    },
  ];
  testTable.forEach((t) => {
    it(t.name, () => {
      expect(search(undefined, t.action)).toMatchSnapshot();
    });
  });
});
