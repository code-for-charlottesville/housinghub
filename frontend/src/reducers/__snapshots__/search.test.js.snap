// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducers/appState set pagination query 1`] = `
Object {
  "errors": "",
  "query": Object {
    "pagination": Object {
      "page": 3,
      "results_per_page": 15,
    },
    "searchFields": Object {},
  },
  "searchResults": Object {
    "pagination": Object {
      "page": 0,
      "results_per_page": 0,
      "totalNumberOfResults": 0,
    },
    "results": Array [],
  },
}
`;

exports[`reducers/appState set pagination results 1`] = `
Object {
  "errors": "",
  "query": Object {
    "pagination": Object {
      "page": 0,
      "results_per_page": 15,
    },
    "searchFields": Object {},
  },
  "searchResults": Object {
    "pagination": Object {
      "page": 5,
      "results_per_page": 0,
      "totalNumberOfResults": 0,
    },
    "results": Array [],
  },
}
`;

exports[`reducers/appState set query results 1`] = `
Object {
  "errors": "",
  "query": Object {
    "pagination": Object {
      "page": 0,
      "results_per_page": 15,
    },
    "searchFields": Object {},
  },
  "searchResults": undefined,
}
`;

exports[`reducers/appState set search fields query 1`] = `
Object {
  "errors": "",
  "query": Object {
    "pagination": Object {
      "page": 0,
      "results_per_page": 15,
    },
    "searchFields": Object {
      "bathrooms": 2,
    },
  },
  "searchResults": Object {
    "pagination": Object {
      "page": 0,
      "results_per_page": 0,
      "totalNumberOfResults": 0,
    },
    "results": Array [],
  },
}
`;
